src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,79 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = clonedWheelList(wheels);
+        this.engine = engine != null ? engine.clone() : null;
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return clonedWheelList(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine != null ? engine.clone() : null;
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(this.year, this.color, new ArrayList<>(wheels), engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(this.year, newColor, new ArrayList<>(wheels), this.engine);
```

Comment:
```suggestion
        return new Car(year, newColor, new ArrayList<>(wheels), engine);
```
You are cloning your wheels in Car's constructor. Anyway, will the ArrayList constructor produces a deep clone of the list?

Reply:
No, it won't make the deep clone, Array lists are removed now from methods